dune/actions/spice.py

```diff
@@ -96,30 +85,333 @@ def _execute(self, game_state):
                 if c != "Shai-Hulud":
                     previous_space = c
                     break
 
-            new_game_state.shai_hulud = previous_space
-            new_game_state.round_state = nexus.NexusRound()
+            _shai_hulud(game_state, previous_space)
+    else:
+        space = game_state.map_state[card]
+        if game_state.storm_position != space.spice_sector:
+            space.spice = space.spice_amount
+        game_state.round_state.drawn_card = True
+        game_state.round_state.stage = "fremen-worm-karama"
+
+    game_state.spice_discard.insert(0, card)
+
+
+class PassFremenWormKarama(Action):
+    name = "pass-fremen-worm-karama"
+    ck_round = "spice"
+    ck_stage = "fremen-worm-karama"
+    ck_faction = "fremen"
+
+    @classmethod
+    def _check(cls, game_state, faction):
+        return game_state.turn > 1
+
+    def _execute(self, game_state):
+        new_game_state = deepcopy(game_state)
+        if new_game_state.round_state.drawn_card:
+            new_game_state.round_state = (nexus.NexusRound()
+                                      if new_game_state.round_state.needs_nexus
+                                      else bidding.BiddingRound())
         else:
-            space = new_game_state.map_state[card]
+            _draw_spice_card(new_game_state)
+        return new_game_state
+
+
+class FremenWormKarama(Action):
+    name = "fremen-worm-karama"
+    ck_round = "spice"
+    ck_stage = "fremen-worm-karama"
+    ck_faction = "fremen"
+    ck_karama = True
+
+    @classmethod
+    def parse_args(cls, faction, args):
+        return FremenWormKarama(faction, args)
+
+    @classmethod
+    def _check(cls, game_state, faction):
+        return game_state.turn > 1
+
+    @classmethod
+    def get_arg_spec(cls, faction=None, game_state=None):
+        return args.Space()
+
+    def __init__(self, faction, space):
+        self.faction = faction
+        self.space = space
+
+    def _execute(self, game_state):
+        new_game_state = deepcopy(game_state)
+        _shai_hulud(new_game_state, self.space)
+        discard_karama(new_game_state, "fremen")
+        return new_game_state
+
+
+class KaramaWorm(Action):
+    name = "karama-worm"
+    ck_round = "spice"
+    ck_stage = "worm"
+    ck_karama = True
+
+    def __init__(self, faction):
+        self.faction = faction
+
+    @classmethod
+    def _check(cls, game_state, faction):
+        if faction == "fremen":
+            raise IllegalAction("You can't karama your own desert powers")
+
+        if faction in game_state.round_state.stage_state.substage_state.karama_passes:
+            raise IllegalAction("You already passed")
+
+    def _execute(self, game_state):
+        new_game_state = deepcopy(game_state)
+        _shai_hulud_eat_forces(new_game_state,
+                               new_game_state.round_state.stage_state.faction_turn)
+        _progress_worm(new_game_state)
+        discard_karama(new_game_state, self.faction)
```

Comment:
I guess this should actually block all future protections/worm rides this round... but that feels super edge casey

Reply:
Yeah, not concerned